cosmos_server/src/structure/systems/laser_cannon_system.rs

```diff
@@ -63,68 +61,124 @@ fn update_system(
-                        server.broadcast_message(
-                            NettyChannelServer::StructureSystems,
-                            cosmos_encoder::serialize(&ServerStructureSystemMessages::LaserCannonSystemFired { ship_entity }),
-                        );
-                    }
-                }
+        else {
+            continue;
+        };
+        let Ok(mut energy_storage_system) = systems.query_mut(&mut es_query) else {
+            continue;
+        };
+
+        let sec = time.elapsed_seconds();
+
+        let mut any_fired = false;
+
+        let default_cooldown = SystemCooldown {
+            cooldown_time: Duration::from_millis(1000),
+            ..Default::default()
+        };
+
+        cooldown.remove_unused_cooldowns(cannon_system);
+
+        for line in cannon_system.lines.iter() {
+            let cooldown = cooldown.lines.entry(line.start.coords()).or_insert(default_cooldown);
+
+            if sec - cooldown.last_use_time < cooldown.cooldown_time.as_secs_f32() {
+                continue;
+            }
+
+            if !((system_active || line.active()) && energy_storage_system.get_energy() >= line.property.energy_per_shot) {
+                continue;
             }
+
+            cooldown.last_use_time = sec;
+            any_fired = true;
+            energy_storage_system.decrease_energy(line.property.energy_per_shot);
+
+            let location = structure.block_world_location(line.start.coords(), global_transform, location);
+
+            let relative_direction = line.direction.to_vec3();
+            let laser_velocity = global_transform.affine().matrix3.mul_vec3(relative_direction) * LASER_BASE_VELOCITY;
+
+            let strength = (5.0 * line.len as f32).powf(1.2);
+            let no_hit = Some(system.structure_entity());
+
+            Laser::spawn(
+                location,
+                laser_velocity,
+                ship_velocity.linvel,
+                strength,
+                no_hit,
+                &time,
+                *physics_world,
+                &mut commands,
+            );
+
+            let color = line.color;
+
+            server.broadcast_message(
+                NettyChannelServer::StructureSystems,
+                cosmos_encoder::serialize(&ServerStructureSystemMessages::CreateLaser {
+                    color,
+                    location,
+                    laser_velocity,
+                    firer_velocity: ship_velocity.linvel,
+                    strength,
+                    no_hit,
+                }),
+            );
+        }
+
+        if any_fired {
+            server.broadcast_message(
+                NettyChannelServer::StructureSystems,
+                cosmos_encoder::serialize(&ServerStructureSystemMessages::LaserCannonSystemFired { ship_entity }),
+            );
+        }
+    }
+}
+
+fn register_logic_connections_for_laser_cannon(blocks: Res<Registry<Block>>, mut registry: ResMut<Registry<LogicBlock>>) {
```

Comment:
To keep the logic system uniform, you might move these logic systems to a file in the specific blocks folder, with the logic indicator and the others.

Reply:
I'm going to keep the laser_cannon_input_event_listener in the laser_cannon_system file, since I think it's very coupled to laser cannon system logic, but I'll move the register_logic_connections_for_laser_cannon.